snuba/datasets/cdc/groupedmessage_processor.py

```diff
@@ -1,54 +1,76 @@
+from datetime import datetime
+from typing import Any, Mapping, Optional
+
+from dataclasses import dataclass
 from dateutil.parser import parse as dateutil_parse
-from snuba.datasets.cdc.cdcprocessors import CdcProcessor
+from snuba.datasets.cdc.cdcprocessors import CdcProcessor, CDCMessageRow
 
 
-class GroupedMessageProcessor(CdcProcessor):
+@dataclass(frozen=True)
+class GroupedMessageRow(CDCMessageRow):
+    offset: int
+    id: int
+    record_deleted: bool
+    status: Optional[int] = None
```

Comment:
When is this `None`? I'm guessing these are for deletions? Should that be commented here? This is also confusing because the `from_*` constructors never actually set the deleted flag, because `_process_delete` circumvents the `from_wal` call, even though it's triggered from the WAL.

It'd be nice if these fields were all aggregated together under a single datatype that was Optional and the fields themselves were required (or not) per the logical schema (not the concrete ClickHouse one.)

Reply:
They are indeed for deletions. 
I will try out your idea but first_release_id and (I think) active_at) are legitimately nullable so they would not be included.

Reply:
> They are indeed for deletions.
> I will try out your idea but first_release_id and (I think) active_at) are legitimately nullable so they would not be included.

I think it'd still make sense to include them in the data/payload/etc structure, but be optional in there. I just think it's a lot clearer denoting "this field was optional because it was deleted and there's no data at all" vs "this field is optional and doesn't have a value", if that makes sense.

This might be too big of a change to make as part of this, and I'm OK with punting it to later if need be. I think we should eventually change the schema itself (again, the "logical" one in the dataset schema, not the concrete generated ClickHouse one) to separate fields the same way so it's clear which fields are nullable for what reason(s). There are a lot of fields that are nullable right now but not nullable in any non-deletion use case and I think that makes reading the schema pretty confusing.

Reply:
done

Reply:
sorry this is not done yet. I commented on the wrong message

Reply:
now it is done